File: app/src/main/java/net/veldor/flibustaloader/http/TorWebClient.java
package net.veldor.flibustaloader.http;

import android.net.Uri;
import android.os.Build;
import android.util.Log;
import android.widget.Toast;

import androidx.documentfile.provider.DocumentFile;

import com.msopentech.thali.android.toronionproxy.AndroidOnionProxyManager;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.MyConnectionSocketFactory;
import net.veldor.flibustaloader.MySSLConnectionSocketFactory;
import net.veldor.flibustaloader.MyWebViewClient;
import net.veldor.flibustaloader.R;
import net.veldor.flibustaloader.database.entity.BooksDownloadSchedule;
import net.veldor.flibustaloader.ecxeptions.BookNotFoundException;
import net.veldor.flibustaloader.ecxeptions.ConnectionLostException;
import net.veldor.flibustaloader.utils.FilesHandler;
import net.veldor.flibustaloader.utils.MyPreferences;
import net.veldor.flibustaloader.utils.URLHelper;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import cz.msebera.android.httpclient.Header;
import cz.msebera.android.httpclient.HttpHeaders;
import cz.msebera.android.httpclient.HttpResponse;
import cz.msebera.android.httpclient.NameValuePair;
import cz.msebera.android.httpclient.NoHttpResponseException;
import cz.msebera.android.httpclient.client.HttpClient;
import cz.msebera.android.httpclient.client.entity.UrlEncodedFormEntity;
import cz.msebera.android.httpclient.client.methods.HttpGet;
import cz.msebera.android.httpclient.client.methods.HttpPost;
import cz.msebera.android.httpclient.client.protocol.HttpClientContext;
import cz.msebera.android.httpclient.config.Registry;
import cz.msebera.android.httpclient.config.RegistryBuilder;
import cz.msebera.android.httpclient.conn.socket.ConnectionSocketFactory;
import cz.msebera.android.httpclient.impl.client.HttpClients;
import cz.msebera.android.httpclient.impl.conn.PoolingHttpClientConnectionManager;
import cz.msebera.android.httpclient.message.BasicNameValuePair;
import cz.msebera.android.httpclient.ssl.SSLContexts;

import static net.veldor.flibustaloader.MyWebViewClient.TOR_NOT_RUNNING_ERROR;

public class TorWebClient {

    public static final String ERROR_DETAILS = "error details";
    private HttpClient mHttpClient;
    private HttpClientContext mContext;

    public TorWebClient() throws ConnectionLostException {
        while (App.getInstance().torInitInProgress) {
            try {
                //noinspection BusyWait
                Thread.sleep(100);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        App.getInstance().torInitInProgress = true;
        // попробую стартовать TOR
        TorStarter starter = new TorStarter();
        App.sTorStartTry = 0;
        while (App.sTorStartTry < 4) {
            // есть три попытки, если все три неудачны- верну ошибку
            if (starter.startTor()) {
                GlobalWebClient.mConnectionState.postValue(GlobalWebClient.CONNECTED);
                App.sTorStartTry = 0;
                break;
            } else {
                App.sTorStartTry++;
            }
        }
        App.getInstance().torInitInProgress = false;
// если счётчик больше 3- не удалось запустить TOR, вызову исключение
        if (App.sTorStartTry > 3) {
            getConnectionError();
        }
        try {
            mHttpClient = getNewHttpClient();
            AndroidOnionProxyManager onionProxyManager = starter.getTor();
            int port = onionProxyManager.getIPv4LocalHostSocksPort();
            InetSocketAddress socksaddr = new InetSocketAddress("127.0.0.1", port);
            mContext = HttpClientContext.create();
            mContext.setAttribute("socks.address", socksaddr);
        } catch (IOException e) {
            e.printStackTrace();
            if (e.getMessage() != null && e.getMessage().equals(TOR_NOT_RUNNING_ERROR)) {

                getConnectionError();
            }
        } catch (RuntimeException e) {
            e.printStackTrace();
            if (e.getMessage() != null && e.getMessage().equals(TOR_NOT_RUNNING_ERROR)) {
                getConnectionError();
            }
        }
    }

    private void getConnectionError() throws ConnectionLostException {
        GlobalWebClient.mConnectionState.postValue(GlobalWebClient.DISCONNECTED);
        throw new ConnectionLostException();
    }


    public String request(String text) {
        if (App.getInstance().useMirror) {
            // TODO заменить зеркало
            Log.d("surprise", "TorWebClient request 128: change mirror");
            text = text.replace("http://flibustahezeous3.onion", "https://flibusta.appspot.com");
        }
        try {
            HttpGet httpGet = new HttpGet(text);
            httpGet.setHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36");
            httpGet.setHeader("X-Compress", "null");
            HttpResponse httpResponse = mHttpClient.execute(httpGet, mContext);
            App.getInstance().mLoadAllStatus.postValue("Данные получены");
            InputStream is;
            is = httpResponse.getEntity().getContent();
            return inputStreamToString(is);
        } catch (IOException e) {
            Log.d("surprise", "TorWebClient request 137: page load error");
            App.getInstance().mLoadAllStatus.postValue("Ошибка загрузки страницы");
            //broadcastTorError(e);
            e.printStackTrace();
        }
        return null;
    }

    public String requestNoMirror(String text) {
        try {
            HttpGet httpGet = new HttpGet(text);
            httpGet.setHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36");
            httpGet.setHeader("X-Compress", "null");
            HttpResponse httpResponse = mHttpClient.execute(httpGet, mContext);
            App.getInstance().mLoadAllStatus.postValue("Данные получены");
            InputStream is;
            is = httpResponse.getEntity().getContent();
            return inputStreamToString(is);
        } catch (IOException e) {
            App.getInstance().mLoadAllStatus.postValue("Ошибка загрузки страницы");
            //broadcastTorError(e);
            e.printStackTrace();
        }
        return null;
    }

    private HttpResponse simpleGetRequest(String url) throws IOException {
        if (App.getInstance().useMirror) {
            // TODO заменить зеркало
            Log.d("surprise", "TorWebClient request 128: change mirror");
            url = url.replace("http://flibustahezeous3.onion", "https://flibusta.appspot.com");
        }
        Log.d("surprise", "TorWebClient simpleGetRequest 128: request " + url);
        HttpGet httpGet = new HttpGet(url);
        httpGet.setHeader("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9");
        httpGet.setHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36");
        httpGet.setHeader("X-Compress", "null");
        return mHttpClient.execute(httpGet, mContext);
    }

    private HttpClient getNewHttpClient() {
        Registry<ConnectionSocketFactory> reg = RegistryBuilder.<ConnectionSocketFactory>create()
                .register("http", new MyConnectionSocketFactory())
                .register("https", new MySSLConnectionSocketFactory(SSLContexts.createSystemDefault()))
                .build();
        PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager(reg, new MyWebViewClient.FakeDnsResolver());
        return HttpClients.custom()
                .setConnectionManager(cm)
                .build();
    }

    private String inputStreamToString(InputStream is) {
        try {
            BufferedReader r = new BufferedReader(new InputStreamReader(is));
            StringBuilder total = new StringBuilder();
            for (String line; (line = r.readLine()) != null; ) {
                total.append(line).append('\n');
            }
            return total.toString();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public void downloadBook(BooksDownloadSchedule book) throws BookNotFoundException {
        try {
            HttpResponse response = simpleGetRequest(URLHelper.getBaseOPDSUrl() + book.link);
            // проверю, что запрос выполнен и файл не пуст. Если это не так- попорбую загрузить книгу с основного домена
            if (response != null && response.getStatusLine().getStatusCode() == 200 && response.getEntity().getContentLength() < 1) {
                boolean result = false;
                // тут может быть загрузка книги без указания длины контента, попробую загрузить
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    try {
                        DocumentFile newFile = FilesHandler.getDownloadFile(book, response);
                        if (newFile != null) {
                            result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, newFile);
                            if(newFile.isFile() && newFile.length() > 0){
                                book.loaded = true;
                            }
                        }
                    } catch (Exception e) {
                        try {
                            File file = FilesHandler.getCompatDownloadFile(book);
                            result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, file);
                            if(file.isFile() && file.length() > 0){
                                book.loaded = true;
                            }
                        } catch (Exception e1) {
                            // скачаю файл просто в папку загрузок
                            File file = FilesHandler.getBaseDownloadFile(book);
                            result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, file);
                            if(file.isFile() && file.length() > 0){
                                book.loaded = true;
                            }
                        }
                    }
                } else {
                    File file = FilesHandler.getCompatDownloadFile(book);
                    result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, file);
                    if(file.isFile() && file.length() > 0){
                        book.loaded = true;
                    }
                }
                if (result) {
                    return;
                }
            }
            if (response == null || response.getStatusLine().getStatusCode() != 200 || response.getEntity().getContentLength() < 1) {
                Log.d("surprise", "TorWebClient downloadBook 225: can't load from main mirror");
                // попробую загрузку с резервного адреса
                response = simpleGetRequest(URLHelper.getFlibustaIsUrl() + book.link);
                if (response != null && response.getStatusLine().getStatusCode() == 200 && response.getEntity().getContentLength() < 1) {
                    boolean result = false;
                    // тут может быть загрузка книги без указания длины контента, попробую загрузить
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        try {
                            DocumentFile newFile = FilesHandler.getDownloadFile(book, response);
                            if (newFile != null) {
                                result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, newFile);
                                if(newFile.isFile() && newFile.length() > 0){
                                    book.loaded = true;
                                }
                            }
                        } catch (Exception e) {
                            try {
                                File file = FilesHandler.getCompatDownloadFile(book);
                                result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, file);

                                if(file.isFile() && file.length() > 0){
                                    book.loaded = true;
                                }
                            } catch (Exception e1) {
                                // скачаю файл просто в папку загрузок
                                File file = FilesHandler.getBaseDownloadFile(book);
                                result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, file);
                                if(file.isFile() && file.length() > 0){
                                    book.loaded = true;
                                }
                            }
                        }
                    } else {
                        File file = FilesHandler.getCompatDownloadFile(book);
                        result = GlobalWebClient.handleBookLoadRequestNoContentLength(response, file);
                        if(file.isFile() && file.length() > 0){
                            book.loaded = true;
                        }
                    }
                    if (result) {
                        return;
                    }
                }
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                if (response != null) {
                    try {
                        DocumentFile newFile = FilesHandler.getDownloadFile(book, response);
                        if (newFile != null) {
                            Log.d("surprise", "TorWebClient downloadBook 276: HERE " + newFile.getName());
                            GlobalWebClient.handleBookLoadRequest(response, newFile);
                            if(newFile.isFile() && newFile.length() > 0){
                                book.loaded = true;
                            }
                        }
                    } catch (Exception e) {
                        try {
                            File file = FilesHandler.getCompatDownloadFile(book);
                            GlobalWebClient.handleBookLoadRequest(response, file);
                            if(file.isFile() && file.length() > 0){
                                book.loaded = true;
                            }
                        } catch (Exception e1) {
                            // скачаю файл просто в папку загрузок
                            File file = FilesHandler.getBaseDownloadFile(book);
                            GlobalWebClient.handleBookLoadRequest(response, file);
                            if(file.isFile() && file.length() > 0){
                                book.loaded = true;
                            }
                        }
                    }
                }
            } else {
                File file = FilesHandler.getCompatDownloadFile(book);
                GlobalWebClient.handleBookLoadRequest(response, file);
                if(file.isFile() && file.length() > 0){
                    book.loaded = true;
                }
            }
        } catch (NoHttpResponseException e) {
            // книга недоступна для скачивания
            throw new BookNotFoundException();
        } catch (IOException e) {
            e.printStackTrace();
            Log.d("surprise", "TorWebClient downloadBook: ошибка при сохранении");
            //throw new TorNotLoadedException();
        }
    }


    public boolean login(Uri uri, String login, String password) throws Exception {
        Log.d("surprise", "TorWebClient login start logging in");
        HttpResponse response;
        UrlEncodedFormEntity params;
        params = get2post(uri, login, password);
        try {
            response = executeRequest("http://flibustahezeous3.onion/node?destination=node", null, params);
            App.getInstance().RequestStatus.postValue(App.getInstance().getString(R.string.response_received_message));
            if (response != null) {
                int status = response.getStatusLine().getStatusCode();
                Log.d("surprise", "TorWebClient login status is " + status);
                // получен ответ, попробую извлечь куку
                Header[] cookies = response.getHeaders("set-cookie");
                if (cookies.length > 1) {
                    StringBuilder cookieValue = new StringBuilder();
                    for (Header c :
                            cookies) {
                        String value = c.getValue();
                        /*if(value.startsWith("PERSISTENT_LOGIN")){
                            cookieValue.append(value.substring(0, value.indexOf(";")));
                        }
                        else */
                        if (value.startsWith("SESS")) {
                            cookieValue.append(value.substring(0, value.indexOf(";")));
                        }
                    }
                    MyPreferences.getInstance().saveLoginCookie(cookieValue.toString());
                    App.getInstance().RequestStatus.postValue(App.getInstance().getString(R.string.success_login_message));
                    return true;
                } else {
                    Log.d("surprise", "TorWebClient login no cookie :(");
                    Log.d("surprise", "TorWebClient.java 230 login: " + response.getEntity().getContent());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
            Log.d("surprise", "TorWebClient login error logging in");
        }
        return false;
    }

    private static UrlEncodedFormEntity get2post(Uri url, String login, String password) {
        Set<String> params = url.getQueryParameterNames();
        if (params.isEmpty()) {
            return null;
        }
        List<NameValuePair> paramsArray = new ArrayList<>();
        paramsArray.add(new BasicNameValuePair("openid_identifier", null));
        paramsArray.add(new BasicNameValuePair("name", login));
        paramsArray.add(new BasicNameValuePair("pass", password));
        paramsArray.add(new BasicNameValuePair("persistent_login", "1"));
        paramsArray.add(new BasicNameValuePair("op", "Вход в систему"));
        paramsArray.add(new BasicNameValuePair("form_build_id", "form-sIt20MHWRjpMKIvxdtHOGqLAa4D2GiBnFIXke7LXv7Y"));
        paramsArray.add(new BasicNameValuePair("form_id", "user_login_block"));
        paramsArray.add(new BasicNameValuePair("return_to", "http://flibustahezeous3.onion/openid/authenticate?destination=node"));
        try {
            return new UrlEncodedFormEntity(paramsArray, "utf8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return null;
    }

    @SuppressWarnings("SameParameterValue")
    private HttpResponse executeRequest(String url, Map<String, String> headers, UrlEncodedFormEntity params) throws Exception {
        try {
            AndroidOnionProxyManager tor = App.getInstance().mLoadedTor.getValue();
            if (tor != null) {
                HttpClient httpClient = getNewHttpClient();
                int port = tor.getIPv4LocalHostSocksPort();
                InetSocketAddress socketAddress = new InetSocketAddress("127.0.0.1", port);
                HttpClientContext clientContext = HttpClientContext.create();
                clientContext.setAttribute("socks.address", socketAddress);

                HttpPost request = new HttpPost(url);

                if (params != null) {
                    request.setEntity(params);
                }
                if (headers != null) {
                    for (Map.Entry<String, String> entry : headers.entrySet()) {
                        request.setHeader(entry.getKey(), entry.getValue());
                    }
                }
                request.setHeader(HttpHeaders.ACCEPT, "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9");
                request.setHeader(HttpHeaders.ACCEPT_ENCODING, "ggzip, deflate");
                request.setHeader(HttpHeaders.ACCEPT_LANGUAGE, "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7");
                request.setHeader(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded");
                request.setHeader(HttpHeaders.CACHE_CONTROL, "no-cache");
                request.setHeader("DNT", "1");
                request.setHeader(HttpHeaders.HOST, "flibustahezeous3.onion");
                request.setHeader("Origin", "http://flibustahezeous3.onion");
                request.setHeader(HttpHeaders.PRAGMA, "no-cache");
                request.setHeader("Proxy-Connection", "keep-ali e");
                request.setHeader("Upgrade-Insecure-Requests", "1");
                request.setHeader(HttpHeaders.USER_AGENT, "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36");
                return httpClient.execute(request, clientContext);
            }
        } catch (RuntimeException e) {
            Toast.makeText(App.getInstance(), "Error request", Toast.LENGTH_LONG).show();
        }
        return null;
    }
}

File: app/src/main/java/net/veldor/flibustaloader/http/TorWebClient.kt
package net.veldor.flibustaloader.http

import android.net.Uri
import android.util.Log
import android.widget.Toast
import cz.msebera.android.httpclient.HttpHeaders
import cz.msebera.android.httpclient.HttpResponse
import cz.msebera.android.httpclient.NameValuePair
import cz.msebera.android.httpclient.client.HttpClient
import cz.msebera.android.httpclient.client.entity.UrlEncodedFormEntity
import cz.msebera.android.httpclient.client.methods.HttpGet
import cz.msebera.android.httpclient.client.methods.HttpPost
import cz.msebera.android.httpclient.client.protocol.HttpClientContext
import cz.msebera.android.httpclient.config.RegistryBuilder
import cz.msebera.android.httpclient.conn.socket.ConnectionSocketFactory
import cz.msebera.android.httpclient.impl.client.HttpClients
import cz.msebera.android.httpclient.impl.conn.PoolingHttpClientConnectionManager
import cz.msebera.android.httpclient.message.BasicNameValuePair
import cz.msebera.android.httpclient.ssl.SSLContexts
import net.veldor.flibustaloader.*
import net.veldor.flibustaloader.MyWebViewClient.FakeDnsResolver
import net.veldor.flibustaloader.ecxeptions.ConnectionLostException
import net.veldor.flibustaloader.utils.PreferencesHandler
import net.veldor.flibustaloader.utils.URLHelper
import net.veldor.flibustaloader.view_models.OPDSViewModel
import java.io.*
import java.net.InetSocketAddress
import java.util.*

class TorWebClient {
    private lateinit var mHttpClient: HttpClient
    private lateinit var mContext: HttpClientContext

    @get:Throws(ConnectionLostException::class)
    private val connectionError: Unit
        get() {
            throw ConnectionLostException()
        }

    fun request(incomingText: String): String? {
        try {
            val httpGet = HttpGet(incomingText)
            httpGet.setHeader(
                "User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
            )
            httpGet.setHeader("X-Compress", "null")
            val httpResponse = mHttpClient.execute(httpGet, mContext)
            val `is`: InputStream = httpResponse.entity.content
            return inputStreamToString(`is`)
        } catch (e: IOException) {
            //broadcastTorError(e);
            e.printStackTrace()
        }
        return null
    }

    fun requestNoMirror(text: String?): String? {
        try {
            val httpGet = HttpGet(text)
            httpGet.setHeader(
                "User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
            )
            httpGet.setHeader("X-Compress", "null")
            val httpResponse = mHttpClient.execute(httpGet, mContext)
            val `is`: InputStream = httpResponse.entity.content
            return inputStreamToString(`is`)
        } catch (e: IOException) {
            //broadcastTorError(e);
            e.printStackTrace()
        }
        return null
    }

    @Throws(java.lang.Exception::class)
    fun directRequest(text: String?): String? {
        Log.d("surprise", "directRequest: make direct request $text")
        val httpGet = HttpGet(text)
        httpGet.setHeader(
            "User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
        )
        httpGet.setHeader("X-Compress", "null")
        val httpResponse = mHttpClient.execute(httpGet, mContext)
        Log.d("surprise", "directRequest:response status is ${httpResponse.statusLine.statusCode}")
        val `is`: InputStream = httpResponse.entity.content
        return inputStreamToString(`is`)
    }

    @Throws(IOException::class)
    fun simpleGetRequest(incomingUrl: String): HttpResponse {
        Log.d("surprise", "simpleGetRequest: load $incomingUrl")
        val httpGet = HttpGet(incomingUrl)
        httpGet.setHeader(
            "Accept",
            "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"
        )
        httpGet.setHeader(
            "User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
        )
        httpGet.setHeader("X-Compress", "null")
        return mHttpClient.execute(httpGet, mContext)
    }

    private val newHttpClient: HttpClient
        get() {
            val reg = RegistryBuilder.create<ConnectionSocketFactory>()
                .register("http", MyConnectionSocketFactory())
                .register("https", MySSLConnectionSocketFactory(SSLContexts.createSystemDefault()))
                .build()
            val cm = PoolingHttpClientConnectionManager(reg, FakeDnsResolver())
            return HttpClients.custom()
                .setConnectionManager(cm)
                .build()
        }

    private fun inputStreamToString(`is`: InputStream): String? {
        try {
            val r = BufferedReader(InputStreamReader(`is`))
            val total = StringBuilder()
            var line: String?
            while (r.readLine().also { line = it } != null) {
                total.append(line).append('\n')
            }
            return total.toString()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    @Throws(Exception::class)
    fun login(uri: Uri, login: String, password: String): Boolean {
        // request main page
        val mainPage = rawRequest(URLHelper.getFlibustaUrl())
        val text = UniversalWebClient().responseToString(mainPage)
        // get form id
        val startIndex = text!!.indexOf("form_build_id") + 19
        val formId = text.subSequence(startIndex, startIndex + 48)
        Log.d("surprise", "loginRequest: form id is $formId")

        val response: HttpResponse?
        val params: UrlEncodedFormEntity? = get2post(uri, login, password, formId.toString())
        try {
            response =
                executeRequest(
                    URLHelper.getFlibustaUrl() + "/node?destination=node",
                    hashMapOf(

                )
                    , params)
            App.instance.requestStatus.postValue(
                App.instance.getString(R.string.response_received_message)
            )
            if (response != null) {
                Log.d("surprise", "login: status is ${response.statusLine.statusCode}")
                val cookies = response.getHeaders("set-cookie")
                if (cookies.size > 1) {
                    val cookieValue = StringBuilder()
                    for (c in cookies) {
                        Log.d("surprise", "login: ${c.name} : ${c.value}")
                        val value = c.value
                        if (value.startsWith("SESS")) {
                            cookieValue.append(value.substring(0, value.indexOf(";")))
                        }
                    }
                    Log.d("surprise", "loginRequest: session cookie is $cookieValue")
                    PreferencesHandler.instance.authCookie = cookieValue.toString()
                    Log.d("surprise", "login: success login")
                    App.instance.requestStatus.postValue(
                        App.instance.getString(R.string.success_login_message)
                    )
                    return true
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return false
    }

    @Throws(Exception::class)
    private fun executeRequest(
        @Suppress("SameParameterValue") url: String,
        @Suppress("SameParameterValue") headers: Map<String, String>?,
        params: UrlEncodedFormEntity?
    ): HttpResponse? {
        try {
            val tor = App.instance.mLoadedTor.value
            if (tor != null) {
                val httpClient = newHttpClient
                val port = tor.iPv4LocalHostSocksPort
                val socketAddress = InetSocketAddress("127.0.0.1", port)
                val clientContext = HttpClientContext.create()
                clientContext.setAttribute("socks.address", socketAddress)
                val request = HttpPost(url)
                if (params != null) {
                    request.entity = params
                }
                if (headers != null) {
                    for ((key, value) in headers) {
                        request.setHeader(key, value)
                    }
                }
                Log.d("surprise", "executeRequest: post on $url")
                return httpClient.execute(request, clientContext)
            }
        } catch (e: RuntimeException) {
            Toast.makeText(App.instance, "Error request", Toast.LENGTH_LONG).show()
        }
        return null
    }

    fun rawRequest(link: String): HttpResponse? {
        OPDSViewModel.currentRequestState.postValue("Отправляю запрос")
        val httpGet = HttpGet(link)
        httpGet.setHeader(
            "User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
        )
        val authCookie = PreferencesHandler.instance.authCookie
        if (authCookie != null) {
            httpGet.setHeader("Cookie", authCookie)
        }
        httpGet.setHeader("X-Compress", "null")
        return mHttpClient.execute(httpGet, mContext)
    }

    companion object {
        const val ERROR_DETAILS = "error details"
        fun get2post(url: Uri, login: String, password: String, formId: String): UrlEncodedFormEntity? {
            val paramsArray: MutableList<NameValuePair> = ArrayList()
            paramsArray.add(BasicNameValuePair("openid_identifier", null))
            paramsArray.add(BasicNameValuePair("name", login))
            paramsArray.add(BasicNameValuePair("pass", password))
            paramsArray.add(BasicNameValuePair("persistent_login", "1"))
            paramsArray.add(BasicNameValuePair("op", "Вход в систему"))
            paramsArray.add(BasicNameValuePair("form_build_id", formId))
            paramsArray.add(BasicNameValuePair("form_id", "user_login_block"))
            paramsArray.add(BasicNameValuePair("return_to",URLHelper.getFlibustaUrl() + "/openid/authenticate?destination=node"))
            try {
                return UrlEncodedFormEntity(paramsArray, "utf8")
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
            }
            return null
        }
    }

    init {
        while (TorStarter.liveTorLaunchState.value == TorStarter.TOR_LAUNCH_IN_PROGRESS) {
            try {
                Thread.sleep(100)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }

        val starter = TorStarter()
        while (TorStarter.torStartTry < 4) {
            // есть три попытки, если все три неудачны- верну ошибку
            if (starter.startTor()) {
                TorStarter.liveTorLaunchState.postValue(TorStarter.TOR_LAUNCH_SUCCESS)
                // обнулю счётчик попыток
                TorStarter.torStartTry = 0
                break
            }
            Log.d("surprise", "doWork: failed tor start try")
            // попытка неудачна, плюсую счётчик попыток
            TorStarter.torStartTry++
        }
        if (TorStarter.liveTorLaunchState.value != TorStarter.TOR_LAUNCH_SUCCESS) {
            TorStarter.liveTorLaunchState.postValue(TorStarter.TOR_LAUNCH_FAILED)
            connectionError
        } else {
            try {
                mHttpClient = newHttpClient
                val onionProxyManager = starter.tor
                val port = onionProxyManager!!.iPv4LocalHostSocksPort
                val socksaddr = InetSocketAddress("127.0.0.1", port)
                mContext = HttpClientContext.create()
                mContext.setAttribute("socks.address", socksaddr)
            } catch (e: IOException) {
                e.printStackTrace()
                if (e.message != null && e.message == MyWebViewClient.TOR_NOT_RUNNING_ERROR) {
                    connectionError
                }
            } catch (e: RuntimeException) {
                e.printStackTrace()
                if (e.message != null && e.message == MyWebViewClient.TOR_NOT_RUNNING_ERROR) {
                    connectionError
                }
            }
        }
    }
}